client/query.go

```diff
@@ -70,14 +46,112 @@ func (c *QueryClient) Validate() error {
 }
 
 // QueryEvents initiates an Insights query, returns a response for parsing
-func (c *QueryClient) QueryEvents(nrqlQuery string) (*QueryResponse, error) {
+func (c *QueryClient) QueryEvents(nrqlQuery string) (response *QueryResponse, err error) {
+	response = &QueryResponse{}
+	err = c.Query(nrqlQuery, response)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// Query initiates an Insights query, with the JSON parsed into 'response'
+func (c *QueryClient) Query(nrqlQuery string, response interface{}) (err error) {
+	if response == nil {
+		return errors.New("go-insights: Invalid query response can not be nil")
+	}
+
+	err = c.generateQueryURL(nrqlQuery)
+	if err != nil {
+		return err
+	}
+
 	c.Logger.Debugf("Querying: %s", nrqlQuery)
-	c.generateQueryURL(nrqlQuery)
+	err = c.queryRequest(response)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// queryRequest makes a NRQL query
+func (c *QueryClient) queryRequest(queryResult interface{}) (err error) {
+	var request *http.Request
+	var response *http.Response
 
-	response, queryErr := c.queryRequest(nrqlQuery)
-	if queryErr != nil {
-		return nil, queryErr
+	if len(c.URL.RawQuery) < 1 {
+		return fmt.Errorf("Query string can not be empty")
 	}
 
-	return response, nil
+	if queryResult == nil {
+		return errors.New("Must have pointer for result")
+	}
+
+	request, err = http.NewRequest("GET", c.URL.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	request.Header.Add("Accept", "application/json")
+	request.Header.Add("X-Query-Key", c.QueryKey)
+
+	client := &http.Client{Timeout: c.RequestTimeout}
+
+	response, err = client.Do(request)
+	if err != nil {
+		err = fmt.Errorf("Failed query request for: %v", err)
+		return
+	}
+	defer func() {
+		respErr := response.Body.Close()
+		if respErr != nil && err == nil {
+			err = respErr // Don't mask previous errors
+		}
+	}()
+
+	if response.StatusCode != http.StatusOK {
+		err = fmt.Errorf("Bad response code: %d", response.StatusCode)
+		return
+	}
+
+	err = c.parseResponse(response, queryResult)
+	if err != nil {
+		err = fmt.Errorf("Failed query: %v", err)
+	}
+
+	return err
+}
+
+// generateQueryURL URL encodes the NRQL
+func (c *QueryClient) generateQueryURL(nrqlQuery string) error {
+	if len(nrqlQuery) < 10 {
```

Comment:
10 isn't completely arbitrary here, though it should be a constant somewhere.  Thought about the smallest legit NRQL query I could come up with ("SHOW eventtypes") which is 15, and cut it down by 1/3rd in case something else comes along shorter...